notebooks/WFC3/point_spread_function/psf_utilities.py

```diff
@@ -0,0 +1,402 @@
+    axes[0].imshow(sci_data, origin='lower', norm=norm, aspect='equal', interpolation='nearest')
+    axes[1].imshow(sci_data - resid, origin='lower', norm=norm, aspect='equal', interpolation='nearest')
+    axes[2].imshow(resid, origin='lower', norm=norm, aspect='equal', interpolation='nearest')
+    axes[0].set_title('Data')
+    axes[1].set_title('Model')
+    axes[2].set_title('Residuals')
+    for x in [0, 1, 2]: 
+        axes[x].set_xlim([xcenter-cutout_size/2, xcenter+cutout_size/2])
+        axes[x].set_ylim([ycenter-cutout_size/2, ycenter+cutout_size/2])
+
+    return figure
+
+
+def download_psf_model(file_path, detector, filter):
+
+    '''
+    Download a PSF model from the WFC3 website and validate the file.
+    '''
+
+    if (detector not in ['WFC3UV', 'WFC3IR', 'ACSWFC']):
+        print('The valid detector options are: WFC3UV, WFC3IR, or ACSWFC')
+
+    psf_name = 'PSFSTD_{}_{}.fits'.format(detector, filter)
+    psf_path = '{}/{}'.format(file_path, psf_name)
+    psf_url = 'https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/wfc3/data-analysis/psf/_documents/'
+
+    # Download the PSF file if it doesn't exist.
+    if not os.path.exists(psf_path):
+        print('Downloading:', psf_url+psf_name)
+        urllib.request.urlretrieve(psf_url+psf_name, psf_path)
+
+    # Copy the PSF file to the working directory.
+    if not os.path.exists(psf_name):
+        print('Copying', psf_name, 'to the current directory.')
+        shutil.copy(psf_path, '.')
+
+    # Confirm that the file can be opened successfully.
+    try:
+        hdul = fits.open(psf_name, ignore_missing_end=True)
+        hdul.close()
+        print('Validation complete, the PSF file is readable.')
+    except IOError:
+        raise IOError('ERROR: Unable to open', psf_name)
+    except Exception as e:
+        raise Exception(e)
+
+    return psf_name
+
+
+def make_cutouts(image, star_ids, xis, yis, rpix, scale_stars=True, sub_pixel=True, show_figs=True, verbose=True):
+
+    '''
+    A function that extracts postage stamps or cutouts of
+    sources from a science image provided an image, list of
+    source IDs, x and y detector coordinates, and several 
+    optional paramters. The scale_stars parameter noramlizes 
+    the peak flux of each source to unity, which is required 
+    when taking the mean or median in subsequent steps. The 
+    subpixel option aligns the cutouts at the subpixel level, 
+    rather than using integer array values, utilizing user 
+    provided x and y coordinates. The show_figs and verbose 
+    options allow users to toggle diagnostic figures and 
+    messages on or off depending on the number of sources.
+    '''
+
+    # Extract a subimage centered on the star.
+    image_array = []
+
+    # Create a loop over all stars.
+    for i in range(len(xis)):
+        
+        # Convert to integer and round to nearest value.
+        xi = int(numpy.rint(xis[i]))
```

Comment:
Hi Fred, thanks for confirming this important point. Indeed, this is the desired behavior because we always want stars that are offset by up to and including half a pixel to be rounded to the same integer array value so they will be properly centered and aligned at the subpixel level. I confirmed the rounding is going the right direction by also calculating the center of mass after the shift, and the centroids (from hst1pass or photuils in these cases) always agree to within better than 0.05 pixels. However, I did not explicitly test with a centroid that is an exact half pixel value (e.g. 19.500000). Considering that hst1pass and photuils centroid to several decimal places this would seem rare, but still important. I think all is well and we can proceed with the code as is in this case.